sdks/go/pkg/beam/registration/emitter.go

```diff
@@ -0,0 +1,173 @@
+	e.et = et
+	return nil
+}
+
+func (e *emit2[T1, T2]) Value() interface{} {
+	return e.invoke
+}
+
+func (e *emit2[T1, T2]) invoke(key T1, val T2) {
+	e.value = exec.FullValue{Windows: e.ws, Timestamp: e.et, Elm: key, Elm2: val}
+	if err := e.n.ProcessElement(e.ctx, &e.value); err != nil {
+		panic(err)
+	}
+}
+
+type emit1WithTimestamp[T any] struct {
+	n exec.ElementProcessor
+
+	ctx   context.Context
+	ws    []typex.Window
+	et    typex.EventTime
+	value exec.FullValue
+}
+
+func (e *emit1WithTimestamp[T]) Init(ctx context.Context, ws []typex.Window, et typex.EventTime) error {
+	e.ctx = ctx
+	e.ws = ws
+	e.et = et
+	return nil
+}
+
+func (e *emit1WithTimestamp[T]) Value() interface{} {
+	return e.invoke
+}
+
+func (e *emit1WithTimestamp[T]) invoke(et typex.EventTime, val T) {
+	e.value = exec.FullValue{Windows: e.ws, Timestamp: et, Elm: val}
+	if err := e.n.ProcessElement(e.ctx, &e.value); err != nil {
+		panic(err)
+	}
+}
+
+type emit2WithTimestamp[T1, T2 any] struct {
+	n exec.ElementProcessor
+
+	ctx   context.Context
+	ws    []typex.Window
+	et    typex.EventTime
+	value exec.FullValue
+}
+
+func (e *emit2WithTimestamp[T1, T2]) Init(ctx context.Context, ws []typex.Window, et typex.EventTime) error {
+	e.ctx = ctx
+	e.ws = ws
+	e.et = et
+	return nil
+}
+
+func (e *emit2WithTimestamp[T1, T2]) Value() interface{} {
+	return e.invoke
+}
+
+func (e *emit2WithTimestamp[T1, T2]) invoke(et typex.EventTime, key T1, val T2) {
+	e.value = exec.FullValue{Windows: e.ws, Timestamp: et, Elm: key, Elm2: val}
+	if err := e.n.ProcessElement(e.ctx, &e.value); err != nil {
+		panic(err)
+	}
+}
+
+// Emitter1 registers parameters from your DoFn with a
+// signature func(T) and optimizes their execution.
+// This must be done by passing in type parameters of all inputs as constraints,
+// aka: registration.Emitter1[T]()
+func Emitter1[T1 any]() {
+	e := (*func(T1))(nil)
+	registerFunc := func(n exec.ElementProcessor) exec.ReusableEmitter {
+		return &emit1[T1]{n: n}
+	}
+	exec.RegisterEmitter(reflect.TypeOf(e).Elem(), registerFunc)
+}
+
+// Emitter2 registers parameters from your DoFn with a
+// signature func(T1, T2) and optimizes their execution.
+// This must be done by passing in type parameters of all inputs as constraints,
+// aka: registration.Emitter2[T1, T2]()
+func Emitter2[T1, T2 any]() {
+	e := (*func(T1, T2))(nil)
+	registerFunc := func(n exec.ElementProcessor) exec.ReusableEmitter {
+		return &emit2[T1, T2]{n: n}
+	}
+	if reflect.TypeOf(e).Elem().In(0) == typex.EventTimeType {
+		registerFunc = func(n exec.ElementProcessor) exec.ReusableEmitter {
+			return &emit1WithTimestamp[T2]{n: n}
+		}
+	}
+	exec.RegisterEmitter(reflect.TypeOf(e).Elem(), registerFunc)
+}
+
+// Emitter3 registers parameters from your DoFn with a
+// signature func(T1, T2, T3) and optimizes their execution.
+// This must be done by passing in type parameters of all inputs as constraints,
+// aka: registration.Emitter3[T1, T2, T3]()
+func Emitter3[T1, T2, T3 any]() {
```

Comment:
As written this is going to end up confusing for users: We always assume it's a timestamp and it should always be a timestamp, but as written it's implied it could be any 3 types. Either make this actually be an Emitter3, or make it an EventTimeEmitter2. Either way the documentation should be updated to be explicit about the timestamp restriction here.

It may be worthwhile to simply just go with EmitterV emitter KV in these cases instead for the user API, to reduce how much the user would need to think about these cases. Then any EventTime specializations we do can bit EmitterETV and EmitterETKV.

Reply:
I think I prefer just updating the doc comments (and the constraint in this case) - I think that's a reasonably clean experience for users (just pick the function that matches your input arity), and that way we don't end up with 2 separate functions for emitters with 2 inputs. I updated the comments for clarity.